rfcs/20221116-enable-stochastic-rounding.md

```diff
@@ -0,0 +1,335 @@
+-   Perform deterministic rounding on the operand with the rounding direction
+    determined by the given random value.
+
+Although stochastic rounding is probabilistic, the probability can be explicitly
+controlled by choosing an appropriate random number generator. Users might want
+to reproduce the rounding results for debugging or behavioral enforcement
+purposes. This usually requires a reproducible pseudo random number generator,
+whose output is deterministic by a given state. In tensorflow, this generator is
+exposed via the tf.random.generator API. Since the purpose of quantized training
+is to reduce the bits used in calculations, changing the data type after
+rounding is a necessity. Given these considerations, an API is ultimately
+proposed like below:
+
+```
+tf.stochastic_cast(input, dtype, random_seed=None, random_algo=’Philox’)
+```
+
+<table>
+ <tr><td><strong>Args</strong></td><td><strong>Definition</strong></td></tr>                                      
+ <tr><td>input</td><td>The original tensor to be casted</td></tr>                              
+ <tr><td>dtype</td><td>Desired type after casting</td></tr>       
+ <tr><td>random_seed</td><td>Random seed if users need to control the randomness. If None, the system will generate a seed.</td></tr>                              
+ <tr><td>random_algo</td><td> Random number generator algorithms to be used, currently support ‘Philox’ or ‘Threefry’. </td></tr>
+</table>
+
+Returns |                                                                 |
+------- | ---------------------------------------------------------------
+Tensor  | A tensor whose values are casted, with the same shape as input.
+
+The API returns a tensor that contains the rounded values in specified data type
+with the same shape as the operand.
+
+The API asks users to provide random seed and algorithm so that the random
+number sequence can be reproduced thus they can replay the rounding results.
+
+This option encapsulates the random number generation aiming at simplifying API
+calls. Also, doing it this way, we can potentially avoid OOM issues introduced
+by storing and moving large random tensors by optimizing the backend
+implementations. Otherwise, the OOM issue could be inevitable when users need to
+generate random tensors and put them into the API.
+
+However, as more use cases emerge, there might be occasions where custom random
+number generators other than the default need to be provided. Or there might be
+experimental random number generation algorithms to be passed in. Under such
+circumstances, encapsulation might complicate future developments. However, up
+till now we haven’t received any user requests regarding random number
+generation algorithms or custom generators.
+
+### Alternatives Considered
+
+Similar to the proposal, but this option encapsulates random number generation
+inside of the API call and asks users to provide random seed and random number
+generation algorithms.
```

Comment:
This sounds like a description of the proposal above, not the alternative below. Should it not say something like "this option requires the caller to generate a random number in the range [0, 1)".

Reply:
Thanks for catching! revised